src/wp-includes/fonts/font-face/class-wp-font-face.php

```diff
@@ -0,0 +1,439 @@
+	 *             @type string          $font_style              Optional. The font-style property. Default 'normal'.
+	 *             @type string          $font-weight             Optional. The font-weight property. Default '400'.
+	 *             @type string          $font-display            Optional. The font-display property. Default 'fallback'.
+	 *             The following font-face properties are optional and will not be generated if not provided.
+	 *             @type string          $ascent-override         Optional. The ascent-override property.
+	 *             @type string          $descent-override        Optional. The descent-override property.
+	 *             @type string          $font-stretch            Optional. The font-stretch property.
+	 *             @type string          $font-variant            Optional. The font-variant property.
+	 *             @type string          $font-feature-settings   Optional. The font-feature-settings property.
+	 *             @type string          $font-variation-settings Optional. The font-variation-settings property.
+	 *             @type string          $line-gap-override       Optional. The line-gap-override property.
+	 *             @type string          $size-adjust             Optional. The size-adjust property.
+	 *             @type string          $unicode-range           Optional. The unicode-range property.
+	 *         }
+	 *     }
+	 * }
+	 */
+	public function generate_and_print( array $fonts ) {
+		$fonts = $this->validate_fonts( $fonts );
+
+		// Bail out if there are no fonts are given to process.
+		if ( empty( $fonts ) ) {
+			return;
+		}
+
+		printf(
+			$this->get_style_element(),
+			$this->get_css( $fonts )
+		);
+	}
+
+	/**
+	 * Validates each of the font-face properties.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array $fonts The fonts to valid.
+	 * @return array Prepared font-faces organized by provider and font-family.
+	 */
+	private function validate_fonts( array $fonts ) {
+		$validated_fonts = array();
+
+		foreach ( $fonts as $font_faces ) {
+			foreach ( $font_faces as $font_face ) {
+				$font_face = $this->validate_font_face_properties( $font_face );
+				// Skip if failed validation.
+				if ( false === $font_face ) {
+					continue;
+				}
+
+				$validated_fonts[] = $font_face;
+			}
+		}
+
+		return $validated_fonts;
+	}
+
+	/**
+	 * Validates each font-face property.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array $font_face Font face properties to validate.
+	 * @return false|array Validated font-face on success. Else, false.
+	 */
+	private function validate_font_face_properties( array $font_face ) {
+		$font_face = wp_parse_args( $font_face, $this->font_face_property_defaults );
+
+		// Check the font-family.
+		if ( empty( $font_face['font-family'] ) || ! is_string( $font_face['font-family'] ) ) {
+			trigger_error( 'Font font-family must be a non-empty string.' );
```

Comment:
This change comes from @azaozz, who advise only using `_doing_it_wrong()` for the most severe cases. Since then, `trigger_error()` has been used for things like function/method parameter (input) validation. It exists in Core for things like PHP compat changes.

An architectural proposal for handling input validation (i.e. function/method parameter checks) is in the works which will also include new error handling.

Also noting the code being replaced also used `trigger_error()` for these same validation checks.

Reply:
@hellofromtonya Thank you for the additional context. I may be on board with a new function like `wp_trigger_error()`, but since that is not in place yet, I think the usage here should rather rely on `_doing_it_wrong()`, since in those cases it literally is the developer "doing it wrong" rather than something else leading to the problem.

As you mention, `trigger_error()` is currently mostly used in Core for things that deal with PHP compat, which is not the case here, and calling `trigger_error()` ignores `WP_DEBUG` etc constants which is the main reason I object to this change.

I think we should use `_doing_it_wrong()`, and then later we can update it to use `wp_trigger_error()` once that function has been added to Core (potentially immediately as part of https://core.trac.wordpress.org/ticket/57686).

Reply:
Yea, lets just add `wp_trigger_error()`. I know there are ideas to expand/extend its use in the future, but seems is it needed now :)

Reply:
>Yea, lets just add wp_trigger_error(). I know there are ideas to expand/extend its use in the future, but seems is it needed now :)

Sorry for my delay in responding (been sick). 

Thanks @felixarntz and @azaozz. I think you both make good points.

- I moved [Trac 57686](https://core.trac.wordpress.org/ticket/57686#comment:7) back into the 6.4 milestone to add `wp_trigger_error()`. I agree with you @azaozz that it can be modified in the future when it's time to implement input (function/method parameter) validation.

- For this PR as a compromise, I'll switch `trigger_error()` to `_doing_it_wrong()` with a comment to replace it with `wp_trigger_error()`.

Reply:
Done in commit https://github.com/WordPress/wordpress-develop/pull/5051/commits/e695f6f06de51f304b3686694d0197e2f832813b ✅